### PR TITLE
Add examples on how to access APIs with authentication

### PR DESCRIPTION
**Pull request by cgx027**
Add examples on how to access APIs with authentication.

The content fromhttps://github.com/RackHD/specs/pull/39 re-posted to docs, with httpie changed to cURL.

@RackHD/corecommitters @yyscamper @anhou @WangWinson @iceiilin

### CODE REVIEW

**Comment by WangWinson** [comment]
:+1:  Nice Job!

File: docs/rackhd/authentication.rst [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+    "authUsername": "admin",
+    "authPasswordHash": "KcBN9YobNV0wdux8h0fKNqi4uoKCgGl/j8c6YGlG7iA0PB3P9ojbmANGhDlcSBE0iOTIsYsGbtSsbqP4wvsVcw==",
+    "authPasswordSalt": "zlxkgxjvcFwm0M8sWaGojh25qNYO8tuNWUMN4xKPH93PidwkCAvaX2JItLA3p7BSCWIzkw4GwWuezoMvKf3UXg==",
+    "authTokenSecret": "RackHDRocks!",
+    "authTokenExpireIn": 86400,
+
+Following above configuration, the default username/password for login are set to be admin/admin123,
+and the token will expire 24 hours after it's generated.
+
+Please follow the instruction on `Change the default password`_ if there is a need to change the
```
Comment (yyscamper) [inline docs/rackhd/authentication.rst]: So we will have another document to describe how to change the default password? As I don't see it in this document.

File: docs/rackhd/authentication.rst [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+            "authEnabled": false,
+            "routers": "southbound-api-router"
+        }
+    ],
+
+The first endpoint represents a HTTPs service listening at port 8443 that serves northbound APIs, which are
+APIs that being called by users. Noticed that authEnabled is set to true means the authentication is needed
+for accessing northbound APIs.
+
+The second endpoint represent a HTTP service listening at port 8080 that serves southbound APIs, which are
```
Comment (yyscamper) [inline docs/rackhd/authentication.rst]: Should be `represents`

File: docs/rackhd/authentication.rst [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+With authentication is enabled in config.json configure file for an endpoint,
+see `Setting up HTTP/HTTPS endpoint`_ and `Authentication`_, authentication is needed for
+accessing the API group associated with the specified endpoint.
+
+This section describes how to access APIs that need authentication.
+
+Setup endpoints to enable authentication
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Please refer to `Setting up HTTP/HTTPS endpoint`_ on how to setup endpoints. Simply put,
```
Comment (yyscamper) [inline docs/rackhd/authentication.rst]: The last `,` should be removed

File: docs/rackhd/authentication.rst [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+            "address": "172.31.128.1",
+            "port": 8080,
+            "httpsEnabled": false,
+            "proxiesEnabled": false,
+            "authEnabled": false,
+            "routers": "southbound-api-router"
+        }
+    ],
+
+The first endpoint represents a HTTPs service listening at port 8443 that serves northbound APIs, which are
```
Comment (benbp) [inline docs/rackhd/authentication.rst]: Should be "an HTTPs", you use "an" when followed by hard H or U sounds.

File: docs/rackhd/authentication.rst [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+APIs that being called by users. Noticed that authEnabled is set to true means the authentication is needed
```
Comment (benbp) [inline docs/rackhd/authentication.rst]: Should just be "APIs being called by users", also "Note that authEnabled being set to true means that authentication is needed to access northbound APIs"

File: docs/rackhd/authentication.rst (hunk 2 above)
Comment (benbp) [inline docs/rackhd/authentication.rst]: Also "an HTTP service"

File: docs/rackhd/authentication.rst [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+Copy following items to config.json if they are not there yet.::
+
+    "authUsername": "admin",
+    "authPasswordHash": "KcBN9YobNV0wdux8h0fKNqi4uoKCgGl/j8c6YGlG7iA0PB3P9ojbmANGhDlcSBE0iOTIsYsGbtSsbqP4wvsVcw==",
+    "authPasswordSalt": "zlxkgxjvcFwm0M8sWaGojh25qNYO8tuNWUMN4xKPH93PidwkCAvaX2JItLA3p7BSCWIzkw4GwWuezoMvKf3UXg==",
+    "authTokenSecret": "RackHDRocks!",
+    "authTokenExpireIn": 86400,
+
+Following above configuration, the default username/password for login are set to be admin/admin123,
```
Comment (benbp) [inline docs/rackhd/authentication.rst]: "Following the above configuration"

File: docs/rackhd/authentication.rst (hunk 1 above)
Comment (benbp) [inline docs/rackhd/authentication.rst]: Should be "Please follow the instructions at"

File: docs/rackhd/authentication.rst [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+
+    curl -k -H "Content-Type:application/json" https://localhost:8443/api/1.1/config --header 'authorization: JWT eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VyIjoiYWRtaW4iLCJpYXQiOjE0NTU2MTI5MzMsImV4cCI6MTQ1NTY5OTMzM30.glW-IvWYDBCfDZ6cS_6APoty22PE_Ir5L1mO-YqO3eE-----------' | python -mjson.tool
+      % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
+                                     Dload  Upload   Total   Spent    Left  Speed
+    100    31  100    31    0     0   1806      0 --:--:-- --:--:-- --:--:--  1823
+    {
+        "message": "invalid signature"
+    }
+
+A 401 bad request response with 'No auth token' message will be returned if:
```
Comment (benbp) [inline docs/rackhd/authentication.rst]: "with a"

File: docs/rackhd/authentication.rst (hunk 1 above)
Comment (cgx027) [inline docs/rackhd/authentication.rst]: @yyscamper , yes, a section called "Change the default password" is in the configuration.rst.

**Comment by anhou** [comment]
:+1:

**Comment by cgx027** [comment]
@yyscamper @benbp , PR updated per grammar suggestions.

**Comment by benbp** [comment]
@cgx027 thanks! I have one more comment, but I haven't been able to come up with a valid alternative yet. The sentence below is very hard to understand:

```
With authentication is enabled in config.json configure file for an endpoint,
 see `Setting up HTTP/HTTPS endpoint`_ and `Authentication`_, authentication is needed for
 accessing the API group associated with the specified endpoint.
```

**Comment by benbp** [comment]
Could you perhaps try splitting that sentence up into two? Maybe that would make it more understandable. I'm not quite sure what you are trying to communicate there otherwise I might be able to offer suggestions.

**Comment by cgx027** [comment]
@benbp I was trying to say, if auth is enabled in config.json for a specific endpoint, auth is needed for accessing the APIs defined within that endpoint. I say this in the purpose of reminding user that auth is NOT needed until it's enabled. 
How about this: 
'''
With 'authEnabled' set to 'true' in config.json for an endpoint,
 authentication is needed for accessing the APIs defined within the specified endpoint.
'''

**Comment by cgx027** [comment]
@benbp I was trying to say, if auth is enabled in config.json for a specific endpoint, auth is needed for accessing the APIs defined within that endpoint. I say this in the purpose of reminding user that auth is NOT needed until it's enabled. 
How about this: 

```
With 'authEnabled' set to 'true' in config.json for an endpoint,
authentication is needed for accessing the APIs defined within the specified endpoint.
```

**Comment by benbp** [comment]
Ah ok, let me add to your suggestion then:

```
When 'authEnabled' is set to 'true' in the config.json file for an endpoint, authentication will be needed to access the APIs that are defined within that endpoint.
```

**Comment by benbp** [comment]
Also did you accidentally close the PR?

**Comment by cgx027** [comment]
Seems I did...Just reopen it~~ by the way, your suggestion looks great.

**Comment by benbp** [comment]
:+1: with those changes!

File: docs/rackhd/authentication.rst (hunk 3 above)
Comment (benbp) [inline docs/rackhd/authentication.rst]: I think the comma is valid but optional here.

Comment (cgx027) [inline docs/rackhd/authentication.rst]: Thanks @benbp and @yyscamper . I do intended to leave a pause here, this will break the line shorter.

Comment (yyscamper) [inline docs/rackhd/authentication.rst]: OK for me.

**Comment by yyscamper** [comment]
:+1:

**Pull request merged**